src/app/model/ChessEngine.ts

```diff
@@ -1,3 +1,37 @@
 import { ChessModel } from './ChessModel';
+import { PieceMoves } from './PieceMoves';
+import { PiecePositions } from './PiecesPositions';
+import { PieceWasCaptured, PieceWasMoved, Square } from './Types';
+import { Piece } from './Piece';
+import { Player } from './Player';
+import { ChessBoard } from './ChessBoard';
 
-export class ChessEngine implements ChessModel {}
+export class ChessEngine implements ChessModel, PieceMoves, PiecePositions {
+  move(byPlayer: Player, squareFrom: Square, squareTo: Square, board: ChessBoard): (PieceWasMoved | PieceWasCaptured)[] {
+    const moveEventsList: (PieceWasMoved | PieceWasCaptured)[] = [];
+    const chosenPiece = board.onPositionPiece(squareFrom) as Piece;
+    const possibleMoves = chosenPiece.possibleMoves(squareFrom, board);
+    const isOccupied = board.onPositionPiece(squareTo);
+    const isOponent = isOccupied && isOccupied?.side !== chosenPiece.side;
+
+    if (byPlayer.side === chosenPiece.side) {
+      if (
+        possibleMoves.find((square: Square) => {
+          return square.column == squareTo.column && square.row == squareTo.row;
+        })
+      ) {
+        board.movePiece(squareFrom, squareTo);
+        if (isOponent) moveEventsList.push({ eventType: 'PieceWasCaptured', piece: isOccupied as Piece, onSquare: squareTo });
```

Comment:
To zalozenie jest troszke inne :) Chcemy zwrocic PieceWasMoved ZAWSZE jak sie ruszy, a DODATKOWO PieceWasCaptured jesli jeszcze ten ruch spowoduje zbicie pionka.